### PR TITLE
WIP: Added ImageRange for itk::Image iteration.

### PR DESCRIPTION
**Pull request by N-Dekker**
Supports iterating over the pixels of an itk::Image by a range-based for-loop, e.g.:

     ImageRange<ImageType> range{ *image };

     for (auto&& pixel : range)
     {
       pixel = pixel + 42;
     }

### CODE REVIEW

File: Modules/Core/Common/include/itkImageRange.h [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+      // to avoid AppleClang 6.0.0.6000056 compilation error, "no viable conversion..."
+      m_PixelAccessor(arg.m_PixelAccessor)
+    {
+    }
+
+
+    /**  Returns a reference to the current pixel. */
+    reference operator*() const ITK_NOEXCEPT
+    {
+      return m_PixelAccessor.Get(*m_CurrentPixel);
```
Comment (N-Dekker) [inline Modules/Core/Common/include/itkImageRange.h]: @blowekamp Can you please check if I'm using the _pixel accessor_ correctly, here? The accessor is from image->GetPixelAccessor(), is that the right one?

Comment (N-Dekker) [reply Modules/Core/Common/include/itkImageRange.h]: See also my Discourse post ["PixelAccessor or NeighborhoodAccessor?" ](https://discourse.itk.org/t/pixelaccessor-or-neighborhoodaccessor/1441 )

**Comment by N-Dekker** [comment]
@thewtex Is the name "ImageRange" OK to you? It supports iteration over _the range of pixels of an image buffer_. So maybe it should be named "ImageBufferPixelRange" instead... what do you think?

**Comment by thewtex** [comment]
> @thewtex Is the name "ImageRange" OK to you? It supports iteration over _the range of pixels of an image buffer_. So maybe it should be named "ImageBufferPixelRange" instead... what do you think?

@N-Dekker yes, I think you hit the mark with "ImageRange". "ImageBufferPixelRange" is a good idea, too, but I think the "BufferPixel" is sufficiently implicit and keeping the length down is beneficial.

It bring up another common use, though -- many times this is likely used where a subset of the Buffered `itk::ImageRegion` is desired. Could a second initialization argument be added to specify the target `ImageRegion`?

**Comment by N-Dekker** [comment]
@thewtex I'd have to check the run-time performance cost of adding an extra `itk::ImageRegion` argument to ImageRange. If it is significant, I'd rather have a separate `ImageRegionRange` type.  (Although some code could be shared between that `ImageRegionRange` and the existing `ImageRange`).

**Comment by N-Dekker** [comment]
@thewtex This pull request now consists of three separate WIP commits. Do you think it's technically possible for me to squash them together, eventually, _before_ one of you guys would merge it to the master?

**Comment by N-Dekker** [comment]
> @N-Dekker Yes, please do squash so that we have a unified commit.

@thewtex  I want squash, but I don't know how! Is it an option in GitHub, or is it something I should do at the Git Bash command line? If so, do you know how?

**Comment by thewtex** [comment]
Ah, to squash, use `git rebase -i`, then select  `squash` on the desired commits. Follow the instructions in the terminal, then `git review-push --force`. :tada: :taco:

**Comment by hjmjohnson** [comment]
@N-Dekker "git rebase -i origin/master" will bring up your favorite editor.   The instructions at the bottom of the message describe how to re-order/squash/fixup/reword commits.

**Comment by thewtex** [comment]
More details, e.g. here:

  https://blog.carbonfive.com/2017/08/28/always-squash-and-rebase-your-git-commits/

**Comment by dzenanz** [comment]
If you use TortoiseGit, select the last few commits, right click and choose "Combine into one commit".

**Comment by N-Dekker** [comment]
Thanks to you all @dzenanz @hjmjohnson @thewtex Tomorrow I think I have some time to give it try.

**Comment by N-Dekker** [comment]
Sorry, I did not yet squash. because I keep getting (apparently) unrelated compilation errors (something about itkSymmetricEigenAnalysis.h: no template named 'vnl_matrix_fixed'). At ITK.macOS:
```
2018-11-29T12:33:02.2169870Z In file included from /Users/vsts/agent/2.142.1/work/1/s/Modules/Core/Common/include/itkSymmetricEigenAnalysis.hxx:21:
2018-11-29T12:33:02.2170890Z /Users/vsts/agent/2.142.1/work/1/s/Modules/Core/Common/include/itkSymmetricEigenAnalysis.h:32:52: error: no template named 'vnl_matrix_fixed'
2018-11-29T12:33:02.2170990Z     const TValueType* GetPointerToMatrixData(const vnl_matrix_fixed<TValueType, NRows, NCols> & inputMatrix)
```
Do you have any clue what's going on here?

**Comment by dzenanz** [comment]
[Merging Eigen3](https://github.com/InsightSoftwareConsortium/ITK/commit/671fbbe05bd9d89422de4292e9891160bea1690c) broke stuff.

**Comment by N-Dekker** [comment]
> [Merging Eigen3](https://github.com/InsightSoftwareConsortium/ITK/commit/671fbbe05bd9d89422de4292e9891160bea1690c) broke stuff.

Thanks @dzenanz I think I'd better wait until it's fixed!

**Comment by phcerdan** [comment]
That's on me! CI was green, but ... fixing it. :face_with_head_bandage:

**Comment by phcerdan** [comment]
After reading the CI logs, you will find this useful:
Linux:
```cpp
2018-11-29T12:44:26.2321198Z In file included from /home/vsts/work/1/s/Modules/Core/Common/test/itkImageRangeGTest.cxx:20:0:
2018-11-29T12:44:26.2322268Z /home/vsts/work/1/s/Modules/Core/Common/include/itkImageRange.h: In instantiation of 'itk::Experimental::ImageRange<TImage>::PixelProxy<true, TDummy>::PixelProxy(const itk::Experimental::ImageRange<TImage>::PixelProxy<false>&) [with TDummy = void; TImage = itk::Image<unsigned char>]':
2018-11-29T12:44:26.2322563Z /home/vsts/work/1/s/Modules/Core/Common/test/itkImageRangeGTest.cxx:299:59:   required from here
2018-11-29T12:44:26.2323164Z /home/vsts/work/1/s/Modules/Core/Common/include/itkImageRange.h:98:30: warning: 'itk::Experimental::ImageRange<itk::Image<unsigned char> >::PixelProxy<true, void>::m_InternalPixel' will be initialized after [-Wreorder]
2018-11-29T12:44:26.2324082Z      const InternalPixelType& m_InternalPixel;
2018-11-29T12:44:26.2324340Z                               ^
2018-11-29T12:44:26.2370480Z /home/vsts/work/1/s/Modules/Core/Common/include/itkImageRange.h:95:31: warning:   'const AccessorFunctorType itk::Experimental::ImageRange<itk::Image<unsigned char> >::PixelProxy<true, void>::m_AccessorFunctor' [-Wreorder]
2018-11-29T12:44:26.2370850Z      const AccessorFunctorType m_AccessorFunctor;
2018-11-29T12:44:26.2371041Z                                ^
2018-11-29T12:44:26.2371511Z /home/vsts/work/1/s/Modules/Core/Common/include/itkImageRange.h:120:5: warning:   when initialized here [-Wreorder]
2018-11-29T12:44:26.2371783Z      PixelProxy(const PixelProxy<false>& pixelProxy) ITK_NOEXCEPT
```

**Comment by N-Dekker** [comment]
> After reading the CI logs, you will find this useful:
> itkImageRange.h:98:30: warning: 'itk::Experimental::ImageRange<itk::Image<unsigned char> >::PixelProxy<true, void>::m_InternalPixel' will be initialized after [-Wreorder]

Thanks @phcerdan ! That's very helpful! I must admit I don't always find it easy to see what the compiler complains about, when looking at the dev.azure.com CI log. While CDash isn't perfect either (with CDash I actually missed access to the complete log), I found it usually easier to find what's wrong by just a short glance at the CDash page.

**Comment by dzenanz** [comment]
@N-Dekker you can look at CDash too. PRs report to [experimental](https://open.cdash.org/index.php?project=Insight#!#Experimental) section containing `PR218` in the build name, e.g. `Windows_NT-Build382-PR218-ImageRange`

**Comment by N-Dekker** [comment]
Superseded by #252

**Pull request closed without merging**